File: src/main/scala/edu/osu/cse/groenkeb/logic/proof/interfaces/Proof.java
package edu.osu.cse.groenkeb.logic.proof.interfaces;

import com.google.common.collect.ImmutableSet;

public interface Proof
{
  /**
   * @return the set of all premises given at the start of the proof
   */
  ImmutableSet<Premise> getInitialPremises();
  
  /**
   * @return the terminal conclusion of the proof
   */
  Conclusion getConclusion();
}

File: src/main/scala/edu/osu/cse/groenkeb/logic/proof/types/Proof.scala
package edu.osu.cse.groenkeb.logic.proof.types

sealed abstract class Proof(val conclusion: Option[Conclusion], val premises: Seq[Premise])
case class CompleteProof(conc: Conclusion, prems: Seq[Premise]) extends Proof(Option.apply(conc), prems)
case class NullProof(prems: Seq[Premise]) extends Proof(Option.empty[Conclusion], prems)
